### ISSUE TITLE
consult-focus-lines point remains in invisible region

### ISSUE THREAD

**Issue by jdtsmith** [minad/consult#493]
I enjoy consult-focus-lines in the package-list-packages listing.  I've noticed however that if point is in the text to be hidden, it remains there.  So that if you focus to "consult" (for example) and attempt to mark the first item with "i", the item in the full buffer you were on before focus is marked instead.  Perhaps it would be possible to move point to the nearest position outside any invisible overlay?

**Comment by minad** [comment]
I am not sure if we can do anything about this. Emacs should itself move the point outside of invisible overlays. It seems that Emacs has a hard time to cope with 5000 overlays which are produced in the packages buffer, which may lead to the effects you observe.

I should probably add a warning to consult-focus-lines or change it such that it even refuses to work on buffers with more than 1000 lines. I wrote consult-focus-lines before I learned that overlays don't have a particular efficient native implementation.

**Comment by oantolin** [comment]
Maybe consult-focus-lines could merge adjacent overlays (if it doesn't already).

**Comment by minad** [comment]
> Maybe consult-focus-lines could merge adjacent overlays (if it doesn't already).

Hmm yeah. I also thought about this. I hope the complexity will still be somehow acceptable. Anyways a PR for such an improvement would be welcome.

**Comment by jdtsmith** [comment]
I took a look.  It would be a major simplification and efficiency improvement if the filter returned not a list of line-strings, but a vector of line indices `e.g. [1 5 6 7 8 42 118 ...]` which have matched.  Then it would be simple to spin through this (sorted) vector and end the existing + start a new overlay when `(> ind (1+ last_ind))`.  For many long buffers, you'll have I suspect not very many overlays; either you'll match "a lot of lines", resulting in few, short overlays, or "very few lines" resulting in a few long overlays.  And worst case is "every other line", which is still half as many.  For general files, it would be very challenging to come up with an "every other line" matcher. 

But all-completions does not give anything besides the match string itself.  consult-buffer already provides the solution: append a tofu to each line, match against the lines, then strip the tofus from each match to produce your vector of line-indices.  With this type of optimization, I suspect you could just dump the old and re-create new overlays on each update.

**Comment by jdtsmith** [comment]
BTW, it appears `consult--tofu-encode` uses multiple unicode chars to encode numbers greater than 65k, but `consult--tofu-get` just takes the final char of a candidate.  You could inspect the last chars of the string for tofus, or even just standardize on a constant 2 tofu chars, to have 4.3 billion items as a max limit ;).  Note that, e.g.:

```
% wc -l /usr/share/dict/words                                                                
  235886 /usr/share/dict/words
```

**Comment by minad** [comment]
Sounds good. Tofus are not needed, one can use text properties - this avoids the expensive string reallocation when attaching the tofus. Patch welcome.

**Comment by minad** [comment]
> Also, if 0xFFFE is the maximum offset from the starting character in plane b (which it appears to be), you actually want consult--tofu-range to be 0xFFFF (just as for formatting for decimal you count modulo 10 from 0...9, then divide by 10).

https://charbase.com/10fffd-unicode-plane-private-use-last

**Comment by jdtsmith** [comment]
The many-overlays issue was not the actual source of the problem, but I substantially improved the speed along the lines of my comment above (and also fixed this actual issue) in PR #495.
